### ISSUE TITLE
MArray is slower than Array in reduction

### ISSUE THREAD

**Issue by YingboMa** [JuliaArrays/StaticArrays.jl#540]
MWE:
```julia
julia> using BenchmarkTools

julia> a = @MArray rand(4);

julia> @btime sum( $(Array(a)) )
  3.896 ns (0 allocations: 0 bytes)
1.5713808992800153

julia> @btime sum( $a )
  30.021 ns (1 allocation: 16 bytes)
1.5713808992800153
```

**Comment by c42f** [comment]
This is still the case on the latest julia.  Ultimately it may be due to keyword functions (in this case `mapreduce`) not getting inlined.

**Comment by tkoolen** [comment]
Yeah, in some of the Base reduce/mapreduce code, `NamedTuple`s are sometimes passed around as positional argument instead of using keyword arguments, probably for this reason.

https://github.com/JuliaLang/julia/blob/cce5a6b3d7303a4ca74a7f630b452d29b06e5903/base/reduce.jl#L39

**Comment by mateuszbaran** [comment]
Unfortunately, passing keyword arguments around is relatively slow in Julia (it requires an `invoke`). The easiest solution is to define a bunch of methods like
```
@inline sum(a::StaticArray{<:Tuple,T}) where {T} = reduce(+, a)
```
to avoid passing keyword arguments. I can make a patch if you want.

**Comment by c42f** [comment]
> I can make a patch if you want.

Yes please!

Did you read the code @tkoolen pointed to above? It seemed to me that might be a way to avoid some of the cost.

**Comment by mateuszbaran** [comment]
> Did you read the code @tkoolen pointed to above? It seemed to me that might be a way to avoid some of the cost.

Yes, I'm trying to invent a less tedious solution.

**Comment by c42f** [comment]
Fixed in #659 (thanks to @mateuszbaran :-) )
